### PR TITLE
feat: increment statsd counter when services reach daily limit

### PR DESCRIPTION
**Pull request by jimleroyer**
## Before

Review this PR in notification-utils as this PR relies on it:
https://github.com/cds-snc/notification-utils/pull/72

## Issue

https://trello.com/c/62gscLJH/169-graph-the-services-going-over-the-daily-rate-limit

## Changes

Added a statsd metric when daily rate usage of services have been reached.

## Testing 

* Added two tests covering the two scenarios of when the service rate is within and outside the limit.

### CODE REVIEW

File: tests/app/conftest.py
Comment (AntoineAugusti) [inline tests/app/conftest.py:1232]: Wondering if you could reuse the fixture from utils to avoid duplicating this. I think that in other tests StatsD is mocked with `mocker` directly if you want to avoid adding a new fixture

File: app/notifications/validators.py
Comment (AntoineAugusti) [inline app/notifications/validators.py:37]: I'm not convinced by the name. I fee like introducing 2 namespaces `rate.limit` is too much, "rate limit" should be a thing on its own. Maybe `rate_limit.service_daily`? Do we have other counters in the codebase related to rate limit?

**Review by AntoineAugusti — CHANGES_REQUESTED** [review]
Good work 👍  A few things to consider. Please request another review when the utils PR has been merged and tagged and the CI passes here

File: tests/app/conftest.py
Comment (jimleroyer) [reply tests/app/conftest.py:1232]: Thanks, I am not super familiar with these Python test libraries so any hint is appreciated. I'll try to apply the changes.

Comment (jimleroyer) [reply tests/app/conftest.py]: Regarding reusing the fixtures from *utils*: it seems unfortunately that the module shadowing occurring by both `conftest` files in the same `tests` folder will not allow this.

File: app/notifications/validators.py
Comment (jimleroyer) [reply app/notifications/validators.py:37]: Other counters in *api* project so far:

```
'callback.{}.{}'
'callback.sendgrid.{}'
redis.daily_limit_cache_key(service.id)
'callback.sns.{notification_status}'
'callback.ses.{}'
'callback.ses-smtp.{}'
"user-agent.{}"
"clients.ses.error"
"clients.ses.error"
"clients.ses.error"
"clients.ses.success"
"clients.sendgrid.error"
"clients.sendgrid.success"
"clients.sns.error"
"clients.sns.error"
"clients.sns.success"
"clients.sns.error"
"clients.pinpoint.error"
"clients.pinpoint.error"
"clients.pinpoint.error"
"clients.pinpoint.success"
"clients.pinpoint.error"
```

Comment (jimleroyer) [reply app/notifications/validators.py:37]: Ah naming, one of these hard things.. I will make that change. 

Just one comment to make sure we agree on metrics naming and goals around it.

I want to represent hierarchies of rates using many metrics that we'll be able to catch with this annotation. For example, in the same *validators* module, we could have right now `rate.limit.api` for the `check_service_over_api_rate_limit` method.  

We should be careful in how we name these because we can perform aggregate operations on hierarchies of metrics (which the parent-child associations are represented with `.`s), so that all families of business related metrics should be grouped together.

That said, I am not sure the name I got is the right one... most probably not and we'll not get this right the first time. I would say let's review all these names when we start leveraging statsd and metrics dashboard in a robust manner.

File: tests/app/notifications/test_validators.py
Comment (RenDavidDimen) [inline tests/app/notifications/test_validators.py:178]: I really like the set up for these tests. Very easy to follow/understand! 👍

**Review by RenDavidDimen — APPROVED** [review]
Everything seems to check out for me. I really like your test set up 🙂

**Review by AntoineAugusti — APPROVED** [review]
Good to go 🚢 

You're right to point out that we should revisit StatsD metrics to make sure that they are properly named. I cleaned StatsD prefixes recently https://github.com/cds-snc/notification-api/pull/1183 (+ linked PRs).

Could you open a task in the backlog? 🙏

**Comment by jimleroyer** [comment]
[Head over here for the task](https://trello.com/c/FbZPcyp0/268-review-statsd-metrics-in-api-admin-projects) opened in the product backlog.

**Comment by jimleroyer** [comment]
I will wait for [review of this PR](https://github.com/cds-snc/notification-admin/pull/849) in the *admin* counterpart before merging.

**Pull request merged**
